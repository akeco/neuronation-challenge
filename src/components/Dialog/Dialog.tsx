import { FC, useEffect, useRef, ReactNode } from 'react';
import { createPortal } from 'react-dom';

const modalRoot = document.querySelector('#modal-root') as HTMLElement;

type DialogProps = {
  open: boolean;
  title: string;
  children: ReactNode;
  onClose(): void;
};

const Dialog: FC<DialogProps> = ({ open, children, title, onClose }) => {
  const elRef = useRef<HTMLDivElement | null>(null);
  if (!elRef.current) elRef.current = document.createElement('div');

  useEffect(() => {
    const el = elRef.current!;
    modalRoot?.appendChild(el);
    return () => {
      modalRoot?.removeChild(el);
    };
  }, []);

  const wrapper = (
    <div
      className="fixed flex items-center justify-center left-0 top-0 w-full h-full bg-black bg-opacity-70"
      onClick={onClose}>
      <div className="flex flex-col w-[500px] h-[600px] bg-white rounded-md" onClick={(e) => e.stopPropagation()}>
        <h4 className="border-b border-solid font-medium text-xl p-4 mb-2">{title}</h4>
        <div className="overflow-y-scroll pb-2">{children}</div>
      </div>
    </div>
  );

  if (!open) return null;

  return createPortal(wrapper, elRef.current);
};

export default Dialog;
