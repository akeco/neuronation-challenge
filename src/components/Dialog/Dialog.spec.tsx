import { render, screen, fireEvent } from '@testing-library/react';
import Dialog from 'components/Dialog/Dialog';

const DIALOG_TITLE = 'History';
const DIALOG_CONTENT = 'Content';

test('Test Dialog content', () => {
  const { baseElement } = render(
    <div>
      <div id="modal-root"></div>
      <Dialog open={true} title={DIALOG_TITLE} onClose={() => {}}>
        <p>{DIALOG_CONTENT}</p>
      </Dialog>
    </div>
  );

  expect(baseElement).toMatchSnapshot();
});
