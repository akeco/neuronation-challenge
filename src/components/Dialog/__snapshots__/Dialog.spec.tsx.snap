// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Dialog content 1`] = `
<body>
  <div>
    <div>
      <div
        id="modal-root"
      />
    </div>
  </div>
</body>
`;
